### ISSUE TITLE
ANDN optimizations

### ISSUE THREAD

**Issue by anatoli-starostin** [anatoli-starostin/spiky#6]
- [ ] profile carefully
- [ ] optimize all vector loads with asm
- [ ] backward backprop optimization
  - in no detectors case combined with sparse output gradients may flow only through winners and prewinners
- [ ] fully support cuda streams (may be used for antihebb and backprop), see LUT implementation
- [ ] shared runtime context (for firing_buffer), buffer should be pytorch tensor by analogy with LUTs)
- [ ] before_detectors_gradients as pytorch tensors
- [ ] (minor) optimize export
    - deprecate export_input_synapse_weights
    - unpack groups using FiringBuffer (batch_index for neuron_idx, payload for offset)
      - forward export synapses
      - backward export synapses
- [ ] optimize TPB all other the code (pattern: int tpb = min(ANDN_RUNTIME_KERNELS_TPB, n_detectors))
- [ ] r_/w_/rw_ naming
- [ ] check downsweep/upsweep all over the project for 64 bit support
- [ ] refactor densify according to the new warp based approach
- [ ] use cudaEvents and avoid redundant synchronisations
- [ ] avoid updateCounter in firing buffers (try over launch instead)
